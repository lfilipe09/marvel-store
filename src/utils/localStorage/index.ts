import { Auth } from 'types/authTypes'
import { User } from 'types/userTypes'

const APP_KEY = '@bornlogic'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(`${APP_KEY}_${key}`)
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return JSON.parse(data!)
}

export function setStorageItem(key: string, value: Auth | User[]) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(`${APP_KEY}_${key}`, data)
}

export function removeStorageItem(key: string) {
  if (typeof window === 'undefined') return

  window.localStorage.removeItem(`${APP_KEY}_${key}`)
}
