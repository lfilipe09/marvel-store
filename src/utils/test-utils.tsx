import React from 'react'
import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { AuthContext, AuthInitialValues, AuthProps } from '../hooks/useAuth'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/themes/red'

type CustomRenderProps = {
  cartProviderProps?: AuthProps
} & Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = AuthInitialValues,
    ...renderOptions
  }: CustomRenderProps = {}
) =>
  render(
    <ThemeProvider theme={theme}>
      <AuthContext.Provider value={cartProviderProps}>
        {ui}
      </AuthContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
