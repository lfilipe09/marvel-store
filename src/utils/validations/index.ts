import Joi from 'joi'
import { UserForm, UserFormSignIn } from '../../types/userTypes'
import {
  cpf_schema,
  email_schema,
  name_schema,
  password_schema,
  surname_schema
} from './errors'

const fieldsValidations = {
  cpf: cpf_schema,
  email: email_schema,
  name: name_schema,
  password: password_schema,
  surname: surname_schema
}

export type FieldErrors = {
  [key: string]: string
}

function getUserErrors(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

export function UserValidate(values: UserForm) {
  const { cpf, email, name, password, surname } = fieldsValidations
  const schema = Joi.object({ cpf, email, name, password, surname })

  return getUserErrors(schema.validate(values, { abortEarly: false }))
}

export function UserValidateSignin(values: UserFormSignIn) {
  const { email, password } = fieldsValidations
  const schema = Joi.object({ email, password })

  return getUserErrors(schema.validate(values, { abortEarly: false }))
}

function getFieldErrors(objError: Joi.ValidationResult, field: keyof UserForm) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[field] = err.message
    })
  }

  return errors
}

export function FieldValidate(value: string, field: keyof UserForm) {
  return getFieldErrors(
    fieldsValidations[field].validate(value, { abortEarly: false }),
    field
  )
}
