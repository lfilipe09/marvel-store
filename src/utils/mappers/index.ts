import {
  APICharacterComics,
  APIComics,
  APICreatorComics
} from 'types/api/comic-types'
import { convertToSlug } from 'utils/slug'

export const heroBannerMapper = (comicsData: APIComics) => {
  return comicsData.data.results.map((comic) => ({
    imgUrl: comic.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${comic.thumbnail.path}/detail.${comic.thumbnail.extension}`,
    title: comic.title,
    subtitle: `<p>Leia agora <strong>${comic.title}</strong> disponÃ­vel online`,
    buttonLabel: 'Saiba mais',
    buttonLink: `comics/${comic.id}/${convertToSlug(comic.title)}`
  }))
}

export const mainCardsMapper = (comicsData: APIComics) => {
  return comicsData.data.results.map((comic) => ({
    id: comic.id,
    slug: `comics/${comic.id}/${convertToSlug(comic.title)}`,
    imgUrl: comic.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${comic.thumbnail.path}/portrait_fantastic.${comic.thumbnail.extension}`,
    title: comic.title
  }))
}

export const singleComicMapper = (comicsData: APIComics) => {
  return comicsData.data.results.map((comic) => ({
    id: comic.id,
    imgUrl: comic.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${comic.thumbnail.path}/portrait_fantastic.${comic.thumbnail.extension}`,
    title: comic.title,
    description: comic.description,
    variantDescription: comic.variantDescription,
    format: comic.format,
    pageCount: comic.pageCount,
    gallery: comic.images.map((image) => ({
      path: image.path,
      extension: image.extension
    })),
    creators: {
      available: comic.creators.available,
      collectionURI: comic.creators.collectionURI
    },
    characters: {
      available: comic.characters.available,
      collectionURI: comic.characters.collectionURI
    }
  }))
}

export const singleComicCreatorsMapper = (creatorData: APICreatorComics) => {
  return creatorData.data.results.map((creator) => ({
    id: creator.id,
    firstName: creator.firstName,
    lastName: creator.lastName,
    url: creator.urls[0].url
  }))
}

export const singleComicCharacterMapper = (
  characterData: APICharacterComics
) => {
  return characterData.data.results.map((character) => ({
    imgUrl: character.thumbnail.path.includes('image_not_available')
      ? 'https://i.ibb.co/gZk8B85/marvel-empty-Prancheta-1.png'
      : `${character.thumbnail.path}/detail.${character.thumbnail.extension}`,
    name: character.name,
    url: character.urls[0].url
  }))
}
