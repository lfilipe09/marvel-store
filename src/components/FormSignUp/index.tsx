import React from 'react'
import { AtSign, ChevronRight, Hash, Lock, User } from '@styled-icons/feather'
import Button from 'components/Button'
import TextField from 'components/TextField'
import { useState } from 'react'
import { UserForm } from 'types/userTypes'
import { FieldErrors, UserValidate } from 'utils/validations'
import * as S from './styles'

export type FormSignUpProps = {
  onSubmit: (value: UserForm) => void
}

const FormSignUp = ({ onSubmit }: FormSignUpProps) => {
  const [fieldError, setFieldError] = useState<FieldErrors>({})
  const [values, setValues] = useState<UserForm>({
    cpf: '',
    email: '',
    name: '',
    password: '',
    surname: '',
    profile: 'Administrador'
  })

  const handleInput = (field: string, value: string) => {
    setValues((s) => ({ ...s, [field]: value }))
  }

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault()

    const valuesValidate = values
    //delete valuesValidate['profile']
    const errors = UserValidate(valuesValidate)
    if (Object.keys(errors).length) {
      setFieldError(errors)
      return
    }

    onSubmit(values)
    setFieldError({})
  }

  return (
    <S.Wrapper onSubmit={handleSubmit}>
      <S.InputGroup>
        <TextField
          label={'Nome'}
          name={'name'}
          icon={<User size={'1.5rem'} />}
          placeholder={'Primeiro nome'}
          error={fieldError?.name}
          onInputChange={(v) => {
            handleInput('name', v)
          }}
        />
        <TextField
          label={'Sobrenome'}
          name={'surname'}
          icon={<User size={'1.5rem'} />}
          placeholder={'Ãšltimo nome'}
          error={fieldError?.surname}
          onInputChange={(v) => {
            handleInput('surname', v)
          }}
        />
      </S.InputGroup>
      <TextField
        label={'CPF'}
        name={'cpf'}
        icon={<Hash size={'1.5rem'} />}
        placeholder={'000.000.000-00'}
        mask={'999.999.999-99'}
        error={fieldError?.cpf}
        onInputChange={(v) => {
          handleInput(
            'cpf',
            v
              .split('')
              .filter((char) => /^[0-9]*$/.test(char))
              .join('')
          )
        }}
      />
      <TextField
        label={'E-mail'}
        name={'email'}
        icon={<AtSign size={'1.5rem'} />}
        placeholder={'email@example.com'}
        error={fieldError?.email}
        onInputChange={(v) => {
          handleInput('email', v)
        }}
      />
      <TextField
        label={'Senha'}
        name={'password'}
        icon={<Lock size={'1.5rem'} />}
        placeholder={'Insira sua senha'}
        password={true}
        error={fieldError?.password}
        onInputChange={(v) => {
          handleInput('password', v)
        }}
      />
      <S.ButtonWrapper>
        <Button
          disabled={
            !values.cpf ||
            !values.email ||
            !values.name ||
            !values.password ||
            !values.surname
          }
          type={'submit'}
          icon={<ChevronRight strokeWidth={4} width={'1rem'} />}
        >
          Criar uma conta
        </Button>
      </S.ButtonWrapper>
    </S.Wrapper>
  )
}

export default FormSignUp
