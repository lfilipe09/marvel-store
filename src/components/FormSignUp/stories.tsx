import React from 'react'
import { Story, Meta } from '@storybook/react'
import FormSignUp, { FormSignUpProps } from '.'

export default {
  title: 'Forms/FormSignUp',
  component: FormSignUp,
  parameters: {
    layout: 'fullscreen',
    backgrounds: {
      default: 'bornlogic-primary'
    }
  }
} as Meta

export const Default: Story<FormSignUpProps> = (args) => (
  <div style={{ maxWidth: '36rem' }}>
    <FormSignUp {...args} />
  </div>
)

Default.args = {
  onSubmit: (valores) => console.log('form processado, os valores: ', valores)
}
