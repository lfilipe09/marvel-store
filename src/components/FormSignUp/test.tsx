import React from 'react'
import { render, screen } from 'utils/test-utils'

import FormSignUp from '.'

describe('<FormSignUp />', () => {
  it('should render the form', () => {
    const { container } = render(<FormSignUp onSubmit={() => true} />)

    expect(screen.getByText('E-mail')).toBeInTheDocument()
    expect(screen.getByText(/First name/i)).toBeInTheDocument()
    expect(screen.getByText(/Last name/i)).toBeInTheDocument()
    expect(screen.getByText(/CPF/i)).toBeInTheDocument()
    expect(
      screen.getByRole('button', { name: /Create account/i })
    ).toBeInTheDocument()

    expect(container.parentElement).toMatchSnapshot()
  })
})
