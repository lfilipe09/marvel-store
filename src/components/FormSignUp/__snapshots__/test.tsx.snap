// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignUp /> should render the form 1`] = `
<body>
  <div>
    <form
      class="sc-jqUVSM eoyCmG"
    >
      <div
        class="sc-kDDrLX cDvPVG"
      >
        <div
          class="sc-papXJ MALAL"
        >
          <div
            style="display: flex;"
          >
            <div
              class="sc-gsnTZi llMrpL"
            >
              <div
                class="sc-dkzDqf gMgHpB"
              >
                <label
                  class="sc-jSMfEi gqUHrM"
                  for="name"
                >
                  First name
                </label>
                <input
                  class="sc-hKMtZM bwfWnC"
                  id="name"
                  name="name"
                  placeholder="First name"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="sc-gKXOVf hIWWIB"
              >
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL"
                  fill="none"
                  focusable="false"
                  height="1.5rem"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  viewBox="0 0 24 24"
                  width="1.5rem"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20 21v-2a4 4 0 00-4-4H8a4 4 0 00-4 4v2"
                  />
                  <circle
                    cx="12"
                    cy="7"
                    r="4"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-papXJ MALAL"
        >
          <div
            style="display: flex;"
          >
            <div
              class="sc-gsnTZi llMrpL"
            >
              <div
                class="sc-dkzDqf gMgHpB"
              >
                <label
                  class="sc-jSMfEi gqUHrM"
                  for="surname"
                >
                  Last name
                </label>
                <input
                  class="sc-hKMtZM bwfWnC"
                  id="surname"
                  name="surname"
                  placeholder="Last name"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="sc-gKXOVf hIWWIB"
              >
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL"
                  fill="none"
                  focusable="false"
                  height="1.5rem"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  viewBox="0 0 24 24"
                  width="1.5rem"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20 21v-2a4 4 0 00-4-4H8a4 4 0 00-4 4v2"
                  />
                  <circle
                    cx="12"
                    cy="7"
                    r="4"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-papXJ MALAL"
      >
        <div
          style="display: flex;"
        >
          <div
            class="sc-gsnTZi llMrpL"
          >
            <div
              class="sc-dkzDqf gMgHpB"
            >
              <label
                class="sc-jSMfEi gqUHrM"
                for="cpf"
              >
                CPF
              </label>
              <input
                class="sc-eCYdqJ jLCAfP"
                id="cpf"
                name="cpf"
                placeholder="000.000.000-00"
                type="text"
                value=""
              />
            </div>
            <div
              class="sc-gKXOVf hIWWIB"
            >
              <svg
                aria-hidden="true"
                class="StyledIconBase-ea9ulj-0 lbJwfL"
                fill="none"
                focusable="false"
                height="1.5rem"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                viewBox="0 0 24 24"
                width="1.5rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="4"
                  x2="20"
                  y1="9"
                  y2="9"
                />
                <line
                  x1="4"
                  x2="20"
                  y1="15"
                  y2="15"
                />
                <line
                  x1="10"
                  x2="8"
                  y1="3"
                  y2="21"
                />
                <line
                  x1="16"
                  x2="14"
                  y1="3"
                  y2="21"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-papXJ MALAL"
      >
        <div
          style="display: flex;"
        >
          <div
            class="sc-gsnTZi llMrpL"
          >
            <div
              class="sc-dkzDqf gMgHpB"
            >
              <label
                class="sc-jSMfEi gqUHrM"
                for="email"
              >
                E-mail
              </label>
              <input
                class="sc-hKMtZM bwfWnC"
                id="email"
                name="email"
                placeholder="email@example.com"
                type="text"
                value=""
              />
            </div>
            <div
              class="sc-gKXOVf hIWWIB"
            >
              <svg
                aria-hidden="true"
                class="StyledIconBase-ea9ulj-0 lbJwfL"
                fill="none"
                focusable="false"
                height="1.5rem"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                viewBox="0 0 24 24"
                width="1.5rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="4"
                />
                <path
                  d="M16 8v5a3 3 0 006 0v-1a10 10 0 10-3.92 7.94"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-papXJ MALAL"
      >
        <div
          style="display: flex;"
        >
          <div
            class="sc-gsnTZi llMrpL"
          >
            <div
              class="sc-dkzDqf gMgHpB"
            >
              <label
                class="sc-jSMfEi gqUHrM"
                for="password"
              >
                Password
              </label>
              <input
                class="sc-hKMtZM bwfWnC"
                id="password"
                name="password"
                placeholder="type here your password"
                type="password"
                value=""
              />
            </div>
            <div
              class="sc-gKXOVf hIWWIB"
            >
              <svg
                aria-hidden="true"
                class="StyledIconBase-ea9ulj-0 lbJwfL"
                fill="none"
                focusable="false"
                height="1.5rem"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                viewBox="0 0 24 24"
                width="1.5rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="11"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="11"
                />
                <path
                  d="M7 11V7a5 5 0 0110 0v4"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-iqcoie hOkWGO"
      >
        <button
          class="sc-bczRLJ cUwYPI"
          color="black"
          disabled=""
          type="submit"
        >
          <span>
            Create account
          </span>
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 lbJwfL"
            fill="none"
            focusable="false"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="4"
            viewBox="0 0 24 24"
            width="1rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </button>
      </div>
    </form>
  </div>
</body>
`
